oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java

```diff
@@ -81,47 +85,65 @@ public void setAllowFormEncodedBodyParameter(boolean allowFormEncodedBodyParamet
+		if (authorization != null && authorization.startsWith("Bearer")) {
+			final Matcher matcher = AUTHORIZATION_PATTERN.matcher(authorization);
 			if (!matcher.matches()) {
-				BearerTokenError error = new BearerTokenError(BearerTokenErrorCodes.INVALID_TOKEN,
-						HttpStatus.UNAUTHORIZED,
-						"Bearer token is malformed",
-						"https://tools.ietf.org/html/rfc6750#section-3.1");
-				throw new OAuth2AuthenticationException(error);
+				throw new OAuth2AuthenticationException(BEARER_TOKEN_ERROR_MALFORMED);
 			}
-
 			return matcher.group("token");
 		}
 		return null;
 	}
 
-	private static String resolveFromRequestParameters(HttpServletRequest request) {
-		String[] values = request.getParameterValues("access_token");
-		if (values == null || values.length == 0)  {
+	private String resolveFromRequestParameters(final HttpServletRequest request) {
+		// By this point there is either zero or one value
+		final String value = request.getParameter(ACCESS_TOKEN_PARAM_NAME);
+		if (value != null && isParameterTokenSupportedForRequest(request)) {
+			final Matcher matcher = TOKEN_PATTERN.matcher(value);
+			if (!matcher.matches()) {
+				throw new OAuth2AuthenticationException(BEARER_TOKEN_ERROR_MALFORMED);
+			}
+			return value;
+		} else {
 			return null;
 		}
+	}
 
-		if (values.length == 1) {
-			return values[0];
-		}
+	private boolean isParameterTokenSupportedForRequest(final HttpServletRequest request) {
+		return this.allowUriQueryParameter || (this.allowFormEncodedBodyParameter
+				&& MediaType.APPLICATION_FORM_URLENCODED.equals(MediaType.valueOf(request.getContentType()))
+				&& !RequestMethod.GET.equals(RequestMethod.valueOf(request.getMethod())));
+	}
 
-		BearerTokenError error = new BearerTokenError(BearerTokenErrorCodes.INVALID_REQUEST,
-				HttpStatus.BAD_REQUEST,
-				"Found multiple bearer tokens in the request",
-				"https://tools.ietf.org/html/rfc6750#section-3.1");
-		throw new OAuth2AuthenticationException(error);
+	private boolean hasMultipleAccessTokens(final HttpServletRequest request, final String authorizationHeaderToken) {
+		final String[] parameterTokens = request.getParameterValues(ACCESS_TOKEN_PARAM_NAME);
+		final boolean multipleParameterTokensPresent = parameterTokens != null && parameterTokens.length > 1;
+		final boolean bothParameterTokenAndHeaderPresent = parameterTokens != null && authorizationHeaderToken != null;
+		return multipleParameterTokensPresent || bothParameterTokenAndHeaderPresent;
 	}
 
-	private boolean isParameterTokenSupportedForRequest(HttpServletRequest request) {
-		return ((this.allowFormEncodedBodyParameter && "POST".equals(request.getMethod()))
-				|| (this.allowUriQueryParameter && "GET".equals(request.getMethod())));
+	private boolean isTokenInUriQueryParameter(final HttpServletRequest request) {
+		final String queryString = request.getQueryString();
+		return queryString != null && Arrays.stream(queryString.split("&"))
+				.map(param -> param.split("=")[0])
```

Comment:
HTTP parameters are URL encoded. This means that technically a URL of `/message?%74oken=abc123` is valid. Also note that using this invoking `HttpServletRequest.getParameter("token")` returns `abc123` We should ensure that we first decode the parameter names before we compare.

Reply:
Updated as required